ambry-cloud/src/main/java/com/github/ambry/cloud/OnlineOfflineHelixVcrStateModelFactory.java

```diff
@@ -0,0 +1,41 @@
+/**
+ * Copyright 2020 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.cloud;
+
+import org.apache.helix.participant.statemachine.StateModel;
+import org.apache.helix.participant.statemachine.StateModelFactory;
+
+
+/**
+ * A factory for creating {@link OnlineOfflineHelixVcrStateModel}
+ */
+public class OnlineOfflineHelixVcrStateModelFactory extends StateModelFactory<StateModel> {
```

Comment:
May I ask why we need two state model factories?  The code logic looks same with LeaderStandbyHelixVcrStateModelFactory. I think one factory should suffice.

Reply:
Thanks for pointing this out. I have changed the logic to make factory as the argument.

While the code is similar, there are still a few reasons to maintain separate factories:

1. We need two state model factory classes because the `createStateModel` method that is used by helix to instantiate the state model takes only `resourceName` and `partitionName` as arguments. We need the `createStateModel` method to create the state model class based on the config.
2. When we call `HelixStateMachineEngine.registerStateModelFactory`, the code expects to have a separate factory for each statemodel. Check `org.apache.helix.participant.HelixStateMachineEngine::registerStateModelFactory`. We can of course create and register just one factory and still make it work. But creating separate factory seems cleaner based on the helix implementation.
3. Usually maintaining different factories keeps the test code cleaner too. In this case we are able to get away because helix provides us with a way to do integration testing, but in case any test requires us to pass a custom or mock state model, it can just pass it via config without any test related abstraction leaking into the production code.
4. The helix code examples use the same model too. Check `org.apache.helix.examples.BrokerResourceOnlineOfflineStateModelFactory.java` and `org.apache.helix.examples.BootstrapHandler.java`